File: python/data_structures/graph/adjacency_list/graph_list.py
"""邻接表图：DFS / BFS 演示。"""

from __future__ import annotations
from collections import deque


class GraphList:
    def __init__(self, n: int, directed: bool = False) -> None:
        self.n = n
        self.directed = directed
        self.adj: list[list[tuple[int, int]]] = [[] for _ in range(n)]

    def add_edge(self, u: int, v: int, w: int = 1) -> None:
        self.adj[u].append((v, w))
        if not self.directed:
            self.adj[v].append((u, w))

    def dfs_order(self, start: int) -> list[int]:
        seen = [False] * self.n
        out: list[int] = []

        def dfs(u: int) -> None:
            seen[u] = True
            out.append(u)
            for v, _w in self.adj[u]:
                if not seen[v]:
                    dfs(v)

        dfs(start)
        return out

    def bfs_order(self, start: int) -> list[int]:
        seen = [False] * self.n
        q: deque[int] = deque([start])
        seen[start] = True
        out: list[int] = []
        while q:
            u = q.popleft()
            out.append(u)
            for v, _w in self.adj[u]:
                if not seen[v]:
                    seen[v] = True
                    q.append(v)
        return out


if __name__ == "__main__":
    g = GraphList(4, directed=False)
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    assert g.dfs_order(0) == [0, 1, 3, 2]
    assert g.bfs_order(0) == [0, 1, 2, 3]
    print("GraphList OK")

File: cpp/data_structures/graph/adjacency_list/graph_list.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct Graph {
    int n;
    bool directed;
    vector<vector<pair<int, int>>> adj;
    Graph(int n_, bool d = false) : n(n_), directed(d), adj(n_) {}

    void add_edge(int u, int v, int w = 1) {
        adj[u].push_back({v, w});
        if (!directed) adj[v].push_back({u, w});
    }

    vector<int> dfs_order(int start) {
        vector<int> out;
        vector<bool> seen(n, false);
        function<void(int)> dfs = [&](int u) {
            seen[u] = true;
            out.push_back(u);
            for (auto [v, w] : adj[u])
                if (!seen[v]) dfs(v);
        };
        dfs(start);
        return out;
    }

    vector<int> bfs_order(int start) {
        vector<int> out;
        vector<bool> seen(n, false);
        queue<int> q;
        q.push(start);
        seen[start] = true;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            out.push_back(u);
            for (auto [v, w] : adj[u])
                if (!seen[v]) {
                    seen[v] = true;
                    q.push(v);
                }
        }
        return out;
    }
};

int main() {
    Graph g(4, false);
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 3);
    assert(g.dfs_order(0) == vector<int>({0, 1, 3, 2}));
    assert(g.bfs_order(0) == vector<int>({0, 1, 2, 3}));
    cout << "GraphList OK" << endl;
    return 0;
}
